engine/apps/alerts/tasks/notify_user.py

```diff
@@ -170,51 +194,97 @@ def notify_user_task(
                         notification_error_code=UserNotificationPolicyLogRecord.ERROR_NOTIFICATION_POSTING_TO_SLACK_IS_DISABLED,
                     )
                 else:
-                    log_record = UserNotificationPolicyLogRecord(
-                        author=user,
-                        type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_TRIGGERED,
-                        notification_policy=notification_policy,
-                        alert_group=alert_group,
-                        reason=reason,
-                        slack_prevent_posting=prevent_posting_to_thread,
-                        notification_step=notification_policy.step,
-                        notification_channel=notification_policy.notify_by,
+                    # todo: feature flag
+                    user_notification_bundle, _ = UserNotificationBundle.objects.get_or_create(
+                        user=user, important=important, notification_channel=notification_policy.notify_by
                     )
-        if log_record:  # log_record is None if user notification policy step is unspecified
-            # if this is the first notification step, and user hasn't been notified for this alert group - update metric
-            if (
-                log_record.type == UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_TRIGGERED
-                and previous_notification_policy_pk is None
-                and not user.personal_log_records.filter(
-                    type=UserNotificationPolicyLogRecord.TYPE_PERSONAL_NOTIFICATION_TRIGGERED,
-                    alert_group_id=alert_group_pk,
-                ).exists()
-            ):
-                update_metrics_for_user.apply_async((user.id,))
+                    user_notification_bundle = UserNotificationBundle.objects.filter(
+                        pk=user_notification_bundle.pk
+                    ).select_for_update()[0]
```

Comment:
I have some concerns around both this task and the send_bundled_notification locking the bundle, in a flood scenario there will be a lot of contention for that record and each save gets bigger as we append to notification_data.  We can test it out before we make things more complicated though.

Reply:
In a flood scenario send_bundled_notification will be called once in ~2min. It's not so often, I think, probably it wouldn't be a problem, need to test it 🤔 But I agree that we need to add some limitation for notification_data